### PR TITLE
Restructure Ballerina repositories 

### LINKED ISSUE #27763: FileSystemAlreadyExistsException gets thrown for Project API
**Description:**
When the project API gets called via concurrent invocations the following exception gets thrown.

```
2021-01-08 15:08:51.170 ERROR 1 --- [io-9000-exec-10] o.a.c.c.C.[.[.[/].[dispatcherServlet]    : Servlet.service() for servlet [dispatcherServlet] in context with path [] threw exception [Request processing failed  nested exception is java.lang.NullPointerException] with root cause 
 
 at java.base/java.lang.Thread.run(Thread.java:834) ~[na:na] 
 at java.base/java.lang.Thread.run(Thread.java:834) 
 at org.apache.tomcat.util.threads.TaskThread$WrappingRunnable.run(TaskThread.java:61) 
 at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128) 
 at org.apache.tomcat.util.net.SocketProcessorBase.run(SocketProcessorBase.java:49) 
 at org.apache.tomcat.util.net.NioEndpoint$SocketProcessor.doRun(NioEndpoint.java:1468) 
 at org.apache.coyote.AbstractProtocol$ConnectionHandler.process(AbstractProtocol.java:790) 
 at org.apache.coyote.AbstractProcessorLight.process(AbstractProcessorLight.java:66) 
 at org.apache.coyote.http11.Http11Processor.service(Http11Processor.java:803) 
 at org.apache.catalina.connector.CoyoteAdapter.service(CoyoteAdapter.java:342) 
 at org.apache.catalina.core.StandardEngineValve.invoke(StandardEngineValve.java:87) 
 at org.apache.catalina.valves.ErrorReportValve.invoke(ErrorReportValve.java:81) 
 at org.apache.catalina.core.StandardHostValve.invoke(StandardHostValve.java:140) 
 at org.apache.catalina.authenticator.AuthenticatorBase.invoke(AuthenticatorBase.java:496) 
 at org.apache.catalina.core.StandardContextValve.invoke(StandardContextValve.java:96) 
 at org.apache.catalina.core.StandardWrapperValve.invoke(StandardWrapperValve.java:198) 
 at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:166) 
 at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:193) 
 at org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:107) 
 at org.springframework.web.filter.CharacterEncodingFilter.doFilterInternal(CharacterEncodingFilter.java:200) 
 at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:166) 
 at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:193) 
 at org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:107) 
 at org.springframework.boot.actuate.web.trace.servlet.HttpTraceFilter.doFilterInternal(HttpTraceFilter.java:90) 
 at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:166) 
 at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:193) 
 at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628) 
 at io.ballerina.projects.Bootstrap.loadLangLibFromBalr(Bootstrap.java:157) 
 at io.ballerina.projects.internal.repositories.FileSystemRepository.getPackage(FileSystemRepository.java:104) 
 at io.ballerina.projects.balo.BaloProject.loadProject(BaloProject.java:43) 
 at io.ballerina.projects.internal.PackageConfigCreator.createBalrProjectConfig(PackageConfigCreator.java:78) 
 at io.ballerina.projects.internal.BaloFiles.createPackageManifest(BaloFiles.java:145) 
 at java.base/java.nio.file.FileSystems.newFileSystem(FileSystems.java:286) 
 at java.base/java.nio.file.FileSystems.newFileSystem(FileSystems.java:337) 
 at jdk.zipfs/jdk.nio.zipfs.ZipFileSystemProvider.newFileSystem(ZipFileSystemProvider.java:102) 
java.nio.file.FileSystemAlreadyExistsException
2021-01-08 15:08:51 ERROR ActionInvocationParser:50 - Error in compiling module 
 at io.ballerina.projects.internal.environment.BallerinaDistribution.loadLangLibPackages(BallerinaDistribution.java:69) 
 at io.ballerina.projects.Bootstrap.loadLangLibSymbols(Bootstrap.java:129) 
 at io.ballerina.projects.Bootstrap.loadLangLibFromBalr(Bootstrap.java:163) 
 at io.ballerina.projects.internal.environment.DefaultPackageResolver.resolvePackages(DefaultPackageResolver.java:97) 
 at io.ballerina.projects.internal.environment.DefaultPackageResolver.resolvePackages(DefaultPackageResolver.java:79) 
 at io.ballerina.projects.internal.environment.DefaultPackageResolver.resolveFromRepository(DefaultPackageResolver.java:115) 
 at io.ballerina.projects.internal.environment.DefaultPackageResolver.resolveLangLibPackage(DefaultPackageResolver.java:166) 
 at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:231) 
 at javax.servlet.http.HttpServlet.service(HttpServlet.java:742)

```

**Steps to reproduce:**
This may be reporduced by creating two concurrent requests to build a Ballerina project.

**Affected Versions:**
Ballerina SLP8

**OS, DB, other environment details and versions:**
This was tested in a docker container which was built using alpine:3.12 and openjdk11

**Related Issues (optional):**
N/A

**Suggested Labels (optional):**
N/A

**Suggested Assignees (optional):**
N/A

### PR DESCRIPTION
**Pull request by azinneera**
- Support for the custom local repository to resolve dependencies
   - This can be added using `repository = "local"` to the dependency entry in Dependencies.toml
- The file system repositories will hold the extracted bala file as the dependency package distribution
- API changes
  - Modified `ProjectLoader.loadProject` to support loading a bala project as well
  - Modified `ProjectPaths.packageRoot` to support finding the root of a bala project as well
  - Modified `BalaProject.load` to accept an extracted bal directory as well
  - Introduces `project.dependencyFilePath(String org, String module, String filename) returns Path`; 
 
Fixes #27763

## Check List 
- [x] Read the [Contributing Guide](https://github.com/ballerina-platform/ballerina-lang/blob/master/CONTRIBUTING.md)
- [ ] Updated Change Log
- [ ] Checked Tooling Support (#<Issue Number>)
- [ ] Added necessary tests
  - [x] Unit Tests
  - [ ] Spec Conformance Tests
  - [ ] Integration Tests
  - [ ] Ballerina By Example Tests
- [ ] Increased Test Coverage   
- [ ] Added necessary documentation  
  - [ ] API documentation 
  - [ ] Module documentation in Module.md files
  - [ ] Ballerina By Examples

### CODE REVIEW

File: compiler/ballerina-lang/src/main/java/io/ballerina/projects/PackageDescriptor.java [hunk 1]
```diff
@@ -31,18 +34,26 @@
     private final PackageName packageName;
     private final PackageOrg packageOrg;
     private final PackageVersion packageVersion;
+    private final String repository;
```
Comment (sameerajayasoma) [inline compiler/ballerina-lang/src/main/java/io/ballerina/projects/PackageDescriptor.java:37]: This should be a PackageRespository class IMO, and it should have something like DEFAULT or BUILT_IN to indicate that a dependency is compiling from a built-in repository.

File: compiler/ballerina-lang/src/main/java/io/ballerina/projects/PackageResolution.java [hunk 2]
```diff
@@ -402,7 +413,8 @@ void resolve(ImportModuleRequest importModuleRequest, PackageDependencyScope sco
             }
 
             PackageOrg packageOrg = importModuleRequest.packageOrg();
-            List<PackageName> possiblePkgNames = getPossiblePackageNames(importModuleRequest);
+            List<PackageName> possiblePkgNames;
```
Comment (sameerajayasoma) [inline compiler/ballerina-lang/src/main/java/io/ballerina/projects/PackageResolution.java:416]: Do we want to break this line into two here? Looks unnecessary to me.

File: compiler/ballerina-lang/src/main/java/io/ballerina/projects/PackageResolution.java [hunk 3]
```diff
@@ -197,6 +199,15 @@ PackageVersion getVersionFromPackageManifest(PackageOrg requestedPkgOrg, Package
         return null;
     }
 
+    String getRepositoryFromPackageManifest(PackageOrg requestedPkgOrg, PackageName requestedPkgName) {
```
Comment (sameerajayasoma) [inline compiler/ballerina-lang/src/main/java/io/ballerina/projects/PackageResolution.java:202]: Can we update the `getVersionFromPackageManifest` method to return the dependency? Then we don't need to iterate dependencies twice.

File: compiler/ballerina-lang/src/main/java/io/ballerina/projects/PackageResolution.java [hunk 4]
```diff
@@ -476,6 +491,9 @@ private void addPackageToGraph(ResolutionResponse resolutionResponse) {
         }
 
         private List<PackageName> getPossiblePackageNames(ImportModuleRequest importModuleRequest) {
+            if (langLibPackage(importModuleRequest.packageOrg(), importModuleRequest.moduleName())) {
```
Comment (sameerajayasoma) [inline compiler/ballerina-lang/src/main/java/io/ballerina/projects/PackageResolution.java:494]: Can you please explain why we need to special-case langlib modules here?

**Review by sameerajayasoma — CHANGES_REQUESTED** [review]

File: compiler/ballerina-lang/src/main/java/io/ballerina/projects/PackageResolution.java (hunk 2 above)
Comment (azinneera) [reply compiler/ballerina-lang/src/main/java/io/ballerina/projects/PackageResolution.java:416]: Yes, we can remove it.

File: compiler/ballerina-lang/src/main/java/io/ballerina/projects/PackageResolution.java (hunk 4 above)
Comment (azinneera) [reply compiler/ballerina-lang/src/main/java/io/ballerina/projects/PackageResolution.java:494]: I came across an issue where only `lang` was returned as possible package names. Looks like it was due an issue in a different place. I tested this now by removing it and it works.

Comment (azinneera) [reply compiler/ballerina-lang/src/main/java/io/ballerina/projects/PackageResolution.java]: Fixed with b139f6d

File: compiler/ballerina-lang/src/main/java/io/ballerina/projects/PackageResolution.java (hunk 3 above)
Comment (azinneera) [reply compiler/ballerina-lang/src/main/java/io/ballerina/projects/PackageResolution.java]: Fixed with b139f6d

File: compiler/ballerina-lang/src/main/java/io/ballerina/projects/PackageResolution.java (hunk 2 above)
Comment (azinneera) [reply compiler/ballerina-lang/src/main/java/io/ballerina/projects/PackageResolution.java]: Fixed with b139f6d

File: compiler/ballerina-lang/src/main/java/io/ballerina/projects/PackageDescriptor.java (hunk 1 above)
Comment (azinneera) [reply compiler/ballerina-lang/src/main/java/io/ballerina/projects/PackageDescriptor.java:37]: This change is not simple. Hence, I have created the issue #28988  to track the improvement. Will address it with another PR.

**Review by sameerajayasoma — APPROVED** [review]
Reviewed dependency graph resolution-related changes. LGTM

@hevayo @pramodya1994 Please review repository-related chagnes.

File: compiler/ballerina-lang/src/main/java/io/ballerina/projects/Project.java [hunk 5]
```diff
@@ -97,5 +100,52 @@ protected void populateCompilerContext() {
...
+    /**
+     * Returns the path of the specified file from dependency bala.
+     *
+     * @param org package org of the dependency
+     * @param moduleName module name of the dependency
+     * @param filename name of the file to get the path of
+     * @return path of the file
+     * @throws ProjectException if file cannot be found in the dependency package
+     */
+    public Path dependencyFilePath(String org, String moduleName, String filename) {
```
Comment (sameerajayasoma) [inline compiler/ballerina-lang/src/main/java/io/ballerina/projects/Project.java:114]: As per chat with @azinneera, We decided to remove this method from the `Project` class as this is not the correct place.

**Review by sameerajayasoma — CHANGES_REQUESTED** [review]

File: compiler/ballerina-lang/src/main/java/io/ballerina/projects/Project.java (hunk 5 above)
Comment (azinneera) [reply compiler/ballerina-lang/src/main/java/io/ballerina/projects/Project.java]: Removed this method in 5655dd8

**Review by sameerajayasoma — APPROVED** [review]

**Pull request merged**